README.md
# Linux_Shell_Imlementation

history.txt
history
mkdir try try2 try3
mkdir t
ls
      mkdir       ry
      mkdir    try
ls
ls
ls -l
ls -l -h -a
cd demo
ls
cd
ls
exit
cat history
cat history.txt
cat history
cat history.txt
cat history.txt
ls
rm c.txt
ls
mkdir try2
ls
cd try2
ls
cd try
pwd
ls     -l   -h
cd ..
cd ..
ls
cd ..
cd ..
cd ..
ls
cat history.txt
cat -E history.txt
cat -n history.txt
ls
cat -n -b history.txt
cat history
cat history.txt
exit
date
date -R
ls
cd ..
date
dTE
date
date
date -R
date -f
date -r
ls
ls
ls
ls
history
cd ..
cd ..
history
history
ls
mkdir try
ls -l
 ls -l h
history
cd ..
cd ..
cd ..
history
exit
cd /
pwd
cd home
cd arkasarkar
cd Desktop
ls
lx
ls
ls -l
cd ..
ls
cd
ls
rm history.txt
ls
pwd
cat mozilla.pdf
cat >> history.txt
cat >>kikiki history.txt
cat >> ki.txt



exit

echo -n ":sdsds"
echo *
exit
cd
cd /Desktop/CAOS
cd Desktop/CAOS
mkdir try
ls
cd try
cd ..
cd try
cd ~
cd Desktop
cd /
cd
echo "hello"
cd Desktop/CAOS/try
cd
echo -n "hello"
cd Desktop
echo *
cd CAOS
echo *
exit
history
ls
history 
history 5
history
history
pwd
ls
ls -l
myterminal@arka:~/home/arkasarkar/Desktop/CAOS $ ls
a.out  cat.c  date.c  demo1.c  history.txt  ls.c  makefile  mkdir.c  rm.c  try
ls -h
ls -lh
ls
cat del.txt
date
date -d
date -r
date -R
ls
rm del.txt
mkdir del
ls
rm -r del
mkdir hello
ls
mkdir hello2
ls
exit
ls
rm -r hello
rm -r hello2
ls
rm -r try
ls
history
history
echo "sdsf"
pwd
cat history.txt
exit


rm.c
#include<stdio.h>
#include<unistd.h>
#include<stdlib.h>
#include<sys/types.h>
#include<sys/wait.h>
#include <string.h>

void rm(char parsed[1000][1000]){
    int i=0;
    int j = 0;
    int o = 0;
        int s = 1;

    int flag = 0;
    char path[256] = {};
    char options[256] ={};
    while(i < 1000 && parsed[i][0]!='\n'){
        if(flag==0 && strcmp(parsed[i],"rm")==0){
            flag = 1;
        }
        else if(flag==1 && parsed[i][0]=='-'){
            options[0] = '-';
            options[s] = parsed[i][1];
            ++s;
        }
        else if(flag == 1 && parsed[i][0]!='\0'){
            
            while(i < 1000 && j<256 && parsed[i][j]!='\0'){
                path[o] = parsed[i][j];
                ++j;++o;
            }
            j = 0;
        }
        ++i;
    }
    pid_t pid;
    pid = fork();
    if(pid<0){
        printf("%s\n" , "Some error occurred ");
    }
    else if(pid==0 && options[0]=='\0'){
        execl("/bin/rm","rm",path,(char *)0);
    }
    else if(pid==0 && options[0]!='\0'){
        execl("/bin/rm","rm",options,path,(char *)0);
    }
    else
    {
        wait(NULL);
    }
}

cat.c
#include<stdio.h>
#include<unistd.h>
#include<stdlib.h>
#include<sys/types.h>
#include<sys/wait.h>
#include <string.h>

void cat(char parsed[1000][1000]){
    int i=0;
    int j = 0;
    int o = 0;
    int s = 1;
    int flag = 0;
    char path[256] = {};
    char options[256] ={};
    while(i < 1000 && parsed[i][0]!='\n'){
        if(flag==0 && strcmp(parsed[i],"cat")==0){
            flag = 1;
        }
        else if(flag==1 && parsed[i][0]=='-'){
            options[0] = '-';
            options[s] = parsed[i][1];
            ++s;
        }
        else if(flag==1 && parsed[i][0]=='>'){
            options[0] = '>';
            options[s] = parsed[i][1];
            ++s;
        }
        else if(flag == 1 && parsed[i][0]!='\0'){
            while(i < 1000 && j<256 && parsed[i][j]!='\0'){
                path[o] = parsed[i][j];
                ++j;++o;
            }
            j = 0;
        }
        ++i;
    }
    pid_t pid;
    pid = fork();
    if(pid<0){
        printf("%s\n" , "Some error occurred ");
    }
     else if(pid==0 && options[0]=='\0'){
        execl("/bin/cat","cat",path,(char *)0);
    }
    else if(pid==0 && options[0]!='\0'){
        execl("/bin/cat","cat",options,path,(char *)0);
    }
    else
    {
        wait(NULL);
    }
    
}

date.c
#include<stdio.h>
#include<unistd.h>
#include<stdlib.h>
#include<sys/types.h>
#include<sys/wait.h>
#include <string.h>

void date(char parsed[1000][1000]){
    int i=0;
    int j = 0;
    int o = 0;
    int s = 1;

    int flag = 0;
    char path[256] = {};
    char options[256] ={};
    while(i < 1000 && parsed[i][0]!='\n'){
        if(flag==0 && strcmp(parsed[i],"date")==0){
            flag = 1;
        }
        else if(flag==1 && parsed[i][0]=='-'){
            options[0] = '-';
            options[s] = parsed[i][1];
            ++s;
        }
        else if(flag == 1 && parsed[i][0]!='\0'){
            
            while(i < 1000 && j<256 && parsed[i][j]!='\0'){
                path[o] = parsed[i][j];
                ++j;++o;
            }
            j = 0;
        }
        ++i;
    }
    pid_t pid;
    pid = fork();
    if(pid<0){
        printf("%s\n" , "Some error occurred ");
    }
    else if(pid==0 && options[0]=='\0'){
        execl("/bin/date","date",(char *)0);
    }
    else if(pid==0 && options[0]!='\0'){
        execl("/bin/date","date",options,(char *)0);
    }
    else
    {
        wait(NULL);
    }
}

demo1.c
#include<stdio.h>
#include<unistd.h>
#include<stdlib.h>
#include<sys/types.h>
#include<sys/wait.h>
#include <string.h> 
#include <ctype.h>
#include "ls.c"
#include "mkdir.c"
#include "cat.c"
#include "rm.c"
#include "date.c"
void pwd();
void cd(char parsed[1000][1000]);
int parser(char input[1000],char parsed[1000][1000]);
void echo(char input[1000]);
void history(char parsed[1000][1000]);

FILE *fptr;


int main() {
    while(1){
		char cwd[256] = {};
		getcwd(cwd, 256);
		printf("%s%s%s", "myterminal@arka:~",cwd," $ ");
		char input[1000] = {};
		char parsed[1000][1000] = {{}};
		FILE *ptr;
		ptr = fopen("/home/arkasarkar/Desktop/CAOS/history.txt","a");
		fgets(input,1000,stdin);
		int com = parser(input,parsed);
		fputs(input,ptr);
		fclose(ptr);
		int i;
		switch (com)
		{
		case 0:
			printf("%s\n", "command not found ");
		case 1:
			 cd(parsed);				
			break;
		case 2:
			cat(parsed);
			break;
		case 3:
			echo(input);
			break;
		case 4:
			i = 8;
			int flag = 0;
			while(input[i]!='\0' && input[i]!='\n'){
				if(input[i]=='c'){
					flag = 1;
					break;
				}
				else if(input[i]=='d'){
					flag = 2;
					break;
				}
				else;
				++i;
			}
			if(flag==0) history(parsed);
			else if(flag==1) {
				remove("/home/arkasarkar/Desktop/CAOS/history.txt");
				fopen("/home/arkasarkar/Desktop/CAOS/history.txt","w");
			}
			break;
		case 5:
			pwd();
			break;
		case 6:
			exit(0);
			break;
		case 7:
			rm(parsed);
			break;
		case 8:
			ls1(input);
			break;
		case 9:
			mkdir(parsed);
			break;
		case 10:
			date(parsed);
			break;
		default:
			printf("%s\n", "command not found ");
			break;
		}
	 }
}
void history(char parsed[1000][1000]){
	int i=0;
	int flag = 0;
	char n ;
	while(i<1000 && *parsed[i]!='\n'){
		if(flag==0 && strcmp(parsed[i],"history")==0){
			flag = 1;
		}
		// else if(flag==1 && isdigit(parsed[i])){
		// 	n = *parsed[i];
		// 	flag = 2;
		// }
		++i;
	}
	if(flag == 1){
	fptr = fopen("/home/arkasarkar/Desktop/CAOS/history.txt","r");
	char h = fgetc(fptr);
    while (h != EOF)
    {   printf ("%c", h);
        h = fgetc(fptr);
	}
    fclose(fptr);
	}
	if(flag == 2){
		fptr = fopen("/home/arkasarkar/Desktop/CAOS/history.txt","r");
	char h = fgetc(fptr);
    while (n-->0)
    {   printf ("%c", h);
        h = fgetc(fptr);
	}
    fclose(fptr);
	}
	}

void echo(char input[1000]){
	int i=0;
	while(input[i]!='\0' || input[i]!='\n'){
		if(input[i]=='e' && input[i+1]=='c'&& input[i+2] == 'h'&& input[i+3] == 'o') break;
		++i;
	}

	int j = i+5;
	int flag = 0;

	while(j<1000 && (input[j]!='\0' && input[j]!='\n')){
		if(flag==0 && input[j] == '"') break;
		else if(flag == 0 && input[j] == '-'){
			flag = 1;
			
		}
		else if(flag == 0 && input[j] == '*'){
			flag = 4;
			
		}	

		else if(flag==1 && input[j]=='n'){
			flag = 2;
			break;
		}
		else if(flag==1 && input[j]=='e'){
			flag = 3;
			break;
		}
		else;
		++j;
	}
	char string[256] = {};
	int o = 0;
	if(flag == 0){
		while(j<1000 && o<256 &&  (input[j]!='\0'&& input[j]!='\n')){
			if(input[j]!='"' && input[j]!='\n'){
			string[o] = input[j];
			++o;
		}
		++j;
	}
	printf("%s\n", string);
	}

	if(flag==2){
			int first = 0;

		j = j+1;
		while(j<1000 && o<256 &&  (input[j]!='\0' || input[j]!='\n')){
			if(input[j]=='"') first = 1;
			if(first ==1 && input[j]!='"' && input[j]!='\n'){
			string[o] = input[j];
			++o;
		}
		++j;
	}
	printf("%s", string);
	}
	if(flag==4){
		pid_t pid;
	pid = fork();
	if(pid<0){
		printf("%s\n", "Some error occurred");
	}
	else if(pid==0){
		execl("/bin/ls","ls",(char *)0);
	}
	else{
		wait(NULL);
	}
	}


}
void pwd(){
	char cwd[256];
   if (*getcwd(cwd, 256) != '\0') {
       printf("%s%s\n", "current working directory :- " ,cwd);
   }
   else{
	   printf("Error in pwd");
   }
}

void cd(char parsed[1000][1000]){
	int i=0;
	while(i<1000 && *parsed[i]!='\n'){
		if(strcmp(parsed[i],"cd")!=0){
			break;
		}
		++i;
	}
	if(parsed[i][0]=='\0' || parsed[i][0] =='~') chdir("/home/arkasarkar");
	else chdir(parsed[i])	;
	
}
int parser(char input[1000],char parsed[1000][1000]){
	int i=0;
    int j=0;
    int o = 0;
	for(i=0;i<1000;i++){
		 if(input[i]==' ' || input[i]=='\n' ){
            ++j;
            o = 0;
        }
        else{
            parsed[j][o] = input[i];
            ++o;
        }
	}
	i = 0;
	j = 0;
	if(*parsed[0]=='\n') return 0;
	while(i<1000 && *parsed[i]!='\n'){
		if(strcmp(parsed[i],"cd")==0){
			j=1;
			break;
		}
		if(strcmp(parsed[i],"cat")==0){
			j=2;
			break;
		}
		if(strcmp(parsed[i],"echo")==0){
			j=3;
			break;
		}
		if(strcmp(parsed[i],"history")==0){
			j=4;
			break;
		}
		if(strcmp(parsed[i],"pwd")==0){
			j=5;
			break;
		}
		if(strcmp(parsed[i],"exit")==0){
			j=6;
			break;
		}
		if(strcmp(parsed[i],"rm")==0){
			j=7;
			break;
		}
		if(strcmp(parsed[i],"ls")==0){
			j=8;
			break;
		}
		if(strcmp(parsed[i],"mkdir")==0){
			j=9;
			break;
		}
		
		if(strcmp(parsed[i],"date")==0){
			j=10;
			break;
		}

		
		++i;
	}
	return j;

}



